src/main/java/io/stargate/sgv2/jsonapi/service/operation/model/impl/FindOperation.java

```diff
@@ -395,32 +395,38 @@ public Uni<FindResponse> getDocuments(
    * @return
    */
   public ReadDocument getNewDocument() {
-    ObjectNode rootNode = objectMapper().createObjectNode();
+
+    final var rootNode = objectMapper().createObjectNode();
     DocumentId documentId = null;
-    Stack<LogicalExpression> stack = new Stack<>();
+    final var stack = new Stack<LogicalExpression>();
     stack.push(logicalExpression);
+
     while (!stack.empty()) {
-      LogicalExpression currentLogicalExpression = stack.pop();
+      var currentLogicalExpression = stack.pop();
+
       for (ComparisonExpression currentComparisonExpression :
           currentLogicalExpression.comparisonExpressions) {
         for (DBFilterBase filter : currentComparisonExpression.getDbFilters()) {
-          if (filter instanceof DBFilterBase.IDFilter idFilter && idFilter.canAddField()) {
-            documentId = idFilter.values.get(0);
-            rootNode.putIfAbsent(filter.getPath(), filter.asJson(objectMapper().getNodeFactory()));
+          // every filter must be a collection filter, because we are making a new document and we
+          // only do this for docs
+          if (filter instanceof IDCollectionFilter) {
+            IDCollectionFilter idFilter = (IDCollectionFilter) filter;
+            documentId = idFilter.getSingularDocumentId();
+            idFilter
+                .updateForNewDocument(objectMapper().getNodeFactory())
+                .ifPresent(setOperation -> setOperation.updateDocument(rootNode));
+          } else if (filter instanceof CollectionFilter) {
+            CollectionFilter f = (CollectionFilter) filter;
+            f.updateForNewDocument(objectMapper().getNodeFactory())
+                .ifPresent(setOperation -> setOperation.updateDocument(rootNode));
           } else {
-            if (filter.canAddField()) {
-              JsonNode value = filter.asJson(objectMapper().getNodeFactory());
-              if (value != null) {
-                String filterPath = filter.getPath();
-                SetOperation.constructSet(filterPath, value).updateDocument(rootNode);
-              }
-            }
+            throw new RuntimeException(
```

Comment:
Fine for now, but I think we would do better to use `JsonApiException` with `ErrorCode.SERVER_INTERNAL_ERROR` when cleaning things up.
```throw ErrorCode.SERVER_INTERNAL_ERROR.toApiException("Unsupported filter type in getNextDocument: %s", filter.getClass().getName());
```

Reply:
+1

Reply:
For this one, I can see lots of same things across four pr, I think we could improve them later, since we need to think better error code for each. If rush into it, may make our exception worse.

Reply:
I filed new issue #1283 for that, as a separate PR.